### PR TITLE
[Random Generator] Part2: Migrate functional dropout

### PR DESCRIPTION
**Pull request by VertexC**

### CODE REVIEW

File: oneflow/core/functional/impl/dropout_functor.cpp [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+#include "oneflow/user/kernels/random_mask_like_kernel.h"
+
+namespace oneflow {
+namespace one {
+namespace functional {
+
+namespace impl {
+class DropoutFunctor {
+ public:
+  DropoutFunctor() {
```
Comment (hjchen2) [inline oneflow/core/functional/impl/dropout_functor.cpp:36]: 挪到nn_functor中去吧

File: oneflow/core/functional/impl/dropout_functor.cpp [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        CHECK_JUST(one::OpBuilder("dropout").Input("in").Input("mask").Output("out").Build());
+  }
+  Maybe<Tensor> operator()(const std::shared_ptr<one::Tensor>& x, const float& p,
+                           const Optional<one::Generator>& generator) const {
+    MutableAttrMap random_mask_like_attrs;
+    JUST(random_mask_like_attrs.SetAttr<float>("rate", p));
+
+    std::shared_ptr<one::Generator> gen;
+    if (!generator) {
+      gen = JUST(one::Generator::New("auto"));
```
Comment (hjchen2) [inline oneflow/core/functional/impl/dropout_functor.cpp:49]: 不用New，直接用DefaultGenerator，New出来的每次生成的随机数都一样了

File: oneflow/core/functional/impl/dropout_functor.cpp [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    } else {
+      gen = JUST(generator.value());
+    }
+
+    JUST(random_mask_like_attrs.SetAttr<int64_t>("seed", gen->current_seed()));
+    const auto& random_mask_like_state = std::make_shared<RandomMaskLikeKernelState>(gen);
+
+    const auto& mask = JUST(OpInterpUtil::Dispatch<Tensor>(
+        *random_mask_like_op_, {x}, {random_mask_like_attrs, random_mask_like_state}));
```
Comment (hjchen2) [inline oneflow/core/functional/impl/dropout_functor.cpp:58]: 这里我建议显示声明一个OpInterpContext，因为这个OpInterpContext未来可能还需要增加字段，而且我们默认输入可以直接用`{}`，这样比较直观表示输入有哪些，但OpInterpContext直接用`{}`感觉就不大合适。

File: oneflow/core/functional/impl/dropout_functor.cpp (hunk 2 above)
Comment (VertexC) [reply oneflow/core/functional/impl/dropout_functor.cpp:49]: 如果这里是随机的是不是和module里不太一致？
也就是用户不传入generator的情况下，直接用F.dropout是创建初始随机数不一样的的generator ，而用nn.Dropout创建初始随机数是一样的。

Comment (hjchen2) [reply oneflow/core/functional/impl/dropout_functor.cpp]: F.dropout本身不创建generator，F.dropout只接收外部传给它的generator，否则就用全局的generator，这很符合逻辑啊。nn.Dropout和不传递generator直接用F.dropout肯定是不一样的，你要想一样的，那就也创建一个generator，然后再传递到F.dropout中。

File: oneflow/user/kernels/dropout_kernel.cpp [hunk 4]
```diff
@@ -15,7 +15,7 @@ limitations under the License.
 */
 #include "oneflow/core/framework/framework.h"
 #include "oneflow/user/kernels/op_kernel_state_wrapper.h"
-#include "oneflow/user/kernels/random_mask_generator.h"
+#include "oneflow/user/kernels/random_mask_like_kernel.h"
```
Comment (hjchen2) [inline oneflow/user/kernels/dropout_kernel.cpp:18]: random_mask_like_kernel.h 这个头文件需要么

File: oneflow/user/kernels/random_mask_like_kernel.h [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+class RandomMaskLikeKernelState : public user_op::OpKernelState {
+ public:
+  explicit RandomMaskLikeKernelState(const std::shared_ptr<one::Generator>& generator)
+      : generator_(generator) {}
+
+  const std::shared_ptr<one::Generator>& generator() const { return generator_; }
+
+ private:
+  const std::shared_ptr<one::Generator>& generator_;
```
Comment (hjchen2) [inline oneflow/user/kernels/random_mask_like_kernel.h:31]: 不要用引用，kernel的执行和functor是异步的，引用可能会造成提前析构

**Review by hjchen2 — APPROVED** [review]
